### ISSUE TITLE
Emoji in Headers Triggers MD051 as invalid link fragment

### ISSUE THREAD

**Issue by craglin** [DavidAnson/vscode-markdownlint#384]
When creating a markdown document that includes emojis in the headings, the linter flags the generated anchor tags as incomplete links. This is possibly due to incomplete regex not recognizing the unicode emojis included.

- What did you do?
When adding emoji to headings, functional links are flagged as invalid link fragments.

- What did you expect to happen?
No warning should be triggered, as the link is functional when using a unicode character

- What actually happened?
A warning showed up 

- What messages or errors were there?
`MD051/link-fragments: Link fragments should be validmarkdownlint[MD051](https://github.com/DavidAnson/markdownlint/blob/v0.38.0/doc/md051.md)`

- How can the issue be reproduced?
See code snippet below:

```
# My Header

## Table of Contents

1. [Non-Emoji Link](#non-emoji-heading)
2. [Shortcode Heading, Shortcode Link](#:smile:-shortcode-heading)
3. [Shortcode Heading, Truncated Link](#smile-shortcode-heading)
4. [Shortcode Heading, Unicode Link](#😄-shortcode-heading)
5. [Unicode Heading, Shortcode Link](#:smile:-unicode-heading)
6. [Unicode Heading, Unicode Link](#😄-unicode-heading)

## Non-Emoji Heading

There are no issues with this link. It is both functional and does not throw an error.

## :smile: Shortcode Heading

:smile: in shortcode, spelled out like `:smile:`.
The shortcode link (2) does not work.
The unicode (3) version of the link is functional, but throws an error.

## 😄 Unicode Heading

Unicode link is functional and does not throw error.

## Summary

| Link | Functional | Errors Thrown |
| ---- | ---------- | ------------- |
| 1    | True       | False         |
| 2    | False      | True          |
| 3    | False      | False         |
| 4    | True       | True          |
| 5    | False      | True          |
| 6    | True       | True          |
```

- What version were you using?
markdownlint v0.60.0

- What operating system were you using?
Windows 11 Pro

- Additional information
Relevant markdown extensions being used include:
* GitHub Markdown Preview (bierner.github-markdown-preview) 
  * Markdown Preview Github Styling
  * Markdown Emoji
  * Markdown Checkboxes
  * Markdown yaml Preamble
  * Markdown Footnotes
  * Markdown Preview Mermaid Support
* markdownlint

**Comment by DavidAnson** [comment]
Please show the Markdown to reproduce your issue. I tried based on part of your rendered output above and that produces no violations:

https://dlaa.me/markdownlint/#%25m%23%20%3Asmile%3A%20Spelled%20Heading%0A

**Comment by craglin** [comment]
> Or maybe you are complaining about the "generated anchor tags"? If so, please show what those are. A simple example in the demo playground I use above would be the most helpful.

Hi David. Sorry, I meant to show my markdown as code instead of formatted. I've adjusted my code for clarity, and my post to show the revised code. I also copied it to the demo playground you linked. Here is my [markdown example](https://dlaa.me/markdownlint/#%25m%23%20My%20Header%0A%0A%23%23%20Table%20of%20Contents%0A%0A1.%20%5BNon-Emoji%20Link%5D(%23non-emoji-heading)%0A2.%20%5BShortcode%20Heading%2C%20Shortcode%20Link%5D(%23%3Asmile%3A-shortcode-heading)%0A3.%20%5BShortcode%20Heading%2C%20Truncated%20Link%5D(%23smile-shortcode-heading)%0A4.%20%5BShortcode%20Heading%2C%20Unicode%20Link%5D(%23%F0%9F%98%84-shortcode-heading)%0A5.%20%5BUnicode%20Heading%2C%20Shortcode%20Link%5D(%23%3Asmile%3A-unicode-heading)%0A6.%20%5BUnicode%20Heading%2C%20Unicode%20Link%5D(%23%F0%9F%98%84-unicode-heading)%0A%0A%23%23%20Non-Emoji%20Heading%0A%0AThere%20are%20no%20issues%20with%20this%20link.%20It%20is%20both%20functional%20and%20does%20not%20throw%20an%20error.%0A%0A%23%23%20%3Asmile%3A%20Shortcode%20Heading%0A%0A%3Asmile%3A%20in%20shortcode%2C%20spelled%20out%20like%20%60%3Asmile%3A%60.%0AThe%20shortcode%20link%20(2)%20does%20not%20work.%0AThe%20unicode%20(3)%20version%20of%20the%20link%20is%20functional%2C%20but%20throws%20an%20error.%0A%0A%23%23%20%F0%9F%98%84%20Unicode%20Heading%0A%0AUnicode%20link%20is%20functional%20and%20does%20not%20throw%20error.%0A%0A%23%23%20Summary%0A%0A%7C%20Link%20%7C%20Functional%20%7C%20Errors%20Thrown%20%7C%0A%7C%20----%20%7C%20----------%20%7C%20-------------%20%7C%0A%7C%201%20%20%20%20%7C%20True%20%20%20%20%20%20%20%7C%20False%20%20%20%20%20%20%20%20%20%7C%0A%7C%202%20%20%20%20%7C%20False%20%20%20%20%20%20%7C%20True%20%20%20%20%20%20%20%20%20%20%7C%0A%7C%203%20%20%20%20%7C%20False%20%20%20%20%20%20%7C%20False%20%20%20%20%20%20%20%20%20%7C%0A%7C%204%20%20%20%20%7C%20True%20%20%20%20%20%20%20%7C%20True%20%20%20%20%20%20%20%20%20%20%7C%0A%7C%205%20%20%20%20%7C%20False%20%20%20%20%20%20%7C%20True%20%20%20%20%20%20%20%20%20%20%7C%0A%7C%206%20%20%20%20%7C%20True%20%20%20%20%20%20%20%7C%20True%20%20%20%20%20%20%20%20%20%20%7C%0A). 

You can see the link fragment warnings for lines 6, 8, 9, and 10. There are two ways that I'm importing emojis; one is direct unicode copy-paste and the other is using the shortcode. Shortcode breaks the links when included as `:smile:`, whereas in line 7 there is no warning given, but the code is still nonfunctional.

**Comment by DavidAnson** [comment]
Thank you, this is very helpful! I'll have a look soon. In the meantime, if you are curious please see the bottom of the documentation for this rule which explains that it specifically enforces the GitHub algorithm for representing heading links. Every link has just one representation and this rule expects the document to use that. So there can be cases where a link works yet is still reported as invalid by this rule. Those are by design.

**Comment by DavidAnson** [comment]
I've had a look:

1. Correct (per your table)
2. Correct
3. Why do you say this is not functional. The GitHub version of the target is `#smile-shortcode-heading` which is what's used here.
4. As I wrote above, this may work on GitHub (for me, it doesn't render as a link at all), but it does not use the canonical form (see 3 above), so is deliberately reported.
5. Correct
6. The GitHub version of the target is `#-unicode-heading`, so this is deliberate reported as with 4 above.

I don't see anything improper here unless you can help me understand why 3 is not functional.

**Comment by DavidAnson** [comment]
Defining those additional formats would seem to require code and that's not an extension point for this rule. I would recommend disabling the rule for your scenario.

If this is important for your scenario, you could implement a custom rule that allows the variety of formats you desire.
